url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/triggers/CronExpression.java

```diff
@@ -0,0 +1,1571 @@
+      }
+    }
+
+    addToSet(val, end, 0, type);
+    i++;
+    return i;
+  }
+
+  public String getCronExpression() {
+    return cronExpression;
+  }
+
+  public String getExpressionSummary() {
+    StringBuilder buf = new StringBuilder();
+
+    buf.append("seconds: ");
+    buf.append(getExpressionSetSummary(seconds));
+    buf.append("\n");
+    buf.append("minutes: ");
+    buf.append(getExpressionSetSummary(minutes));
+    buf.append("\n");
+    buf.append("hours: ");
+    buf.append(getExpressionSetSummary(hours));
+    buf.append("\n");
+    buf.append("daysOfMonth: ");
+    buf.append(getExpressionSetSummary(daysOfMonth));
+    buf.append("\n");
+    buf.append("months: ");
+    buf.append(getExpressionSetSummary(months));
+    buf.append("\n");
+    buf.append("daysOfWeek: ");
+    buf.append(getExpressionSetSummary(daysOfWeek));
+    buf.append("\n");
+    buf.append("lastdayOfWeek: ");
+    buf.append(lastdayOfWeek);
+    buf.append("\n");
+    buf.append("nearestWeekday: ");
+    buf.append(nearestWeekday);
+    buf.append("\n");
+    buf.append("NthDayOfWeek: ");
+    buf.append(nthdayOfWeek);
+    buf.append("\n");
+    buf.append("lastdayOfMonth: ");
+    buf.append(lastdayOfMonth);
+    buf.append("\n");
+    buf.append("years: ");
+    buf.append(getExpressionSetSummary(years));
+    buf.append("\n");
+
+    return buf.toString();
+  }
+
+  protected String getExpressionSetSummary(java.util.Set<Integer> set) {
+
+    if (set.contains(NO_SPEC)) {
+      return "?";
+    }
+    if (set.contains(ALL_SPEC)) {
+      return "*";
+    }
+
+    StringBuilder buf = new StringBuilder();
+
+    Iterator<Integer> itr = set.iterator();
+    boolean first = true;
+    while (itr.hasNext()) {
+      Integer iVal = itr.next();
+      String val = iVal.toString();
+      if (!first) {
+        buf.append(",");
+      }
+      buf.append(val);
+      first = false;
+    }
+
+    return buf.toString();
+  }
+
+  protected String getExpressionSetSummary(java.util.ArrayList<Integer> list) {
+
+    if (list.contains(NO_SPEC)) {
+      return "?";
+    }
+    if (list.contains(ALL_SPEC)) {
+      return "*";
+    }
+
+    StringBuilder buf = new StringBuilder();
+
+    Iterator<Integer> itr = list.iterator();
+    boolean first = true;
+    while (itr.hasNext()) {
+      Integer iVal = itr.next();
+      String val = iVal.toString();
+      if (!first) {
+        buf.append(",");
+      }
+      buf.append(val);
+      first = false;
+    }
+
+    return buf.toString();
+  }
+
+  protected int skipWhiteSpace(int i, String s) {
+    for (; i < s.length() && (s.charAt(i) == ' ' || s.charAt(i) == '\t'); i++) {
+      ;
+    }
+
+    return i;
+  }
+
+  protected int findNextWhiteSpace(int i, String s) {
+    for (; i < s.length() && (s.charAt(i) != ' ' || s.charAt(i) != '\t'); i++) {
+      ;
+    }
+
+    return i;
+  }
+
+  protected void addToSet(int val, int end, int incr, int type) throws ParseException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 66. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 886-1017 into a separate method.